com.unity.multiplayer.mlapi/Runtime/Core/NetworkUpdateLoop.cs

```diff
@@ -0,0 +1,328 @@
+                        }
+
+                        // insert after `PostLateUpdate.PlayerSendFrameComplete`
+                        subsystems.Insert(indexOfPlayerSendFrameComplete + 1, NetworkPostLateUpdate.CreateLoopSystem());
+                    }
+                    playerLoopSystem.subSystemList = subsystems.ToArray();
+                }
+
+                customPlayerLoop.subSystemList[i] = playerLoopSystem;
+            }
+
+            PlayerLoop.SetPlayerLoop(customPlayerLoop);
+        }
+
+        /// <summary>
+        /// The current network update stage being executed.
+        /// </summary>
+        public static NetworkUpdateStage UpdateStage;
+
+        /// <summary>
+        /// Registers a network update system to be executed in all network update stages.
+        /// </summary>
+        public static void RegisterAllNetworkUpdates(this INetworkUpdateSystem updateSystem)
+        {
+            foreach (NetworkUpdateStage updateStage in Enum.GetValues(typeof(NetworkUpdateStage)))
+            {
+                RegisterNetworkUpdate(updateSystem, updateStage);
+            }
+        }
+
+        /// <summary>
+        /// Registers a network update system to be executed in a specific network update stage.
+        /// </summary>
+        public static void RegisterNetworkUpdate(this INetworkUpdateSystem updateSystem, NetworkUpdateStage updateStage = NetworkUpdateStage.Update)
+        {
+            int updateStageIndex = (int)updateStage;
+            if (!m_UpdateSystem_Sets[updateStageIndex].Contains(updateSystem))
+            {
+                m_UpdateSystem_Sets[updateStageIndex].Add(updateSystem);
+                m_UpdateSystem_Arrays[updateStageIndex] = m_UpdateSystem_Sets[updateStageIndex].ToArray();
+            }
+        }
+
+        /// <summary>
+        /// Unregisters a network update system from all network update stages.
+        /// </summary>
+        public static void UnregisterAllNetworkUpdates(this INetworkUpdateSystem updateSystem)
+        {
+            foreach (NetworkUpdateStage updateStage in Enum.GetValues(typeof(NetworkUpdateStage)))
+            {
+                UnregisterNetworkUpdate(updateSystem, updateStage);
+            }
+        }
+
+        /// <summary>
+        /// Unregisters a network update system from a specific network update stage.
+        /// </summary>
+        public static void UnregisterNetworkUpdate(this INetworkUpdateSystem updateSystem, NetworkUpdateStage updateStage = NetworkUpdateStage.Update)
+        {
+            int updateStageIndex = (int)updateStage;
+            if (m_UpdateSystem_Sets[updateStageIndex].Contains(updateSystem))
+            {
+                m_UpdateSystem_Sets[updateStageIndex].Remove(updateSystem);
+                m_UpdateSystem_Arrays[updateStageIndex] = m_UpdateSystem_Sets[updateStageIndex].ToArray();
+            }
+        }
+
+        private static readonly HashSet<INetworkUpdateSystem>[] m_UpdateSystem_Sets =
```

Comment:
Seems like you could declare this as a dictionary with key of type NetworkUpdateStage and be more maintainable (vs. remembering to maintain the comments, having a potentially missing index)

Reply:
I was mainly concerned about `Dictionary` lookup on every update but apparently, one lookup per update stage group is not looking too bad. I verified my performance concerns with a local benchmark, both index-to-array and dictHash-to-array approaches are looking almost identical in terms of runtime performance (tested with 128, 1k, 4k and 8k objects).

I'll update shortly.